### PR TITLE
Add Freezetime to CCDB Manager/API

### PR DESCRIPTION
**Pull request by njacazio**

### CODE REVIEW

**Comment by Barthelemy** [comment]
Hi, 
Thanks, this is nice. 
I have a couple of comments but they are up to you, the PR could go through as it is (+the missing test).
1. Naming: Wouldn't it be clearer if the name of this parameter and the associated methods referred to "Validity" ? Personally, I would not have guessed what "freezeTime" referred to (others might find it completely clear though). "upperValidityLimit" perhaps ? 
2. There is the counter-part `If-Not-Before`. It defines the other validity boundary. Thus I would be in favour of adding it as well.

File: CCDB/include/CCDB/BasicCCDBManager.h [hunk 1]
```diff
@@ -92,6 +92,15 @@ class BasicCCDBManager
     }
   }
 
+  /// set the object validity limit
+  void setFreezeTime(long v) { mFreezeTime = v; }
+
+  /// get the object validity limit
+  long getFreezeTime() { return mFreezeTime; }
```
Comment (sawenzel) [inline CCDB/include/CCDB/BasicCCDBManager.h:99]: getter should be marked const

**Comment by njacazio** [comment]
@Barthelemy I agree with your changes, indeed I think it would be more clear this way.
For the `If-Not-Before` I can add the field as well if you find it useful.
@shahor02  I wonder if the line 125 (https://github.com/AliceO2Group/AliceO2/pull/3775/files#diff-3d82a935396c98479328a85f57c8c4a0R125) should also be changed, as if running in non-cached mode asking to limit the object validity would have no effect and I think it might be a confusing feature.

File: CCDB/include/CCDB/BasicCCDBManager.h (hunk 1 above)
Comment (njacazio) [reply CCDB/include/CCDB/BasicCCDBManager.h]: Thank you Sandro, done!

**Comment by shahor02** [comment]
@njacazio for the line https://github.com/AliceO2Group/AliceO2/pull/3775/files#diff-3d82a935396c98479328a85f57c8c4a0R125: Why would you change it? The "time machine" sets the boundaries of the objects creation time by not  their validity. Within these boundaries the you may still have different objects valid for different times, so the caching should work as usual. Have I misunderstood something?

File: CCDB/include/CCDB/BasicCCDBManager.h [hunk 2]
```diff
@@ -92,6 +92,15 @@ class BasicCCDBManager
     }
   }
 
+  /// set the object upper validity limit
+  void setUpperValidity(long v) { mUpperValidity = v; }
```
Comment (shahor02) [inline CCDB/include/CCDB/BasicCCDBManager.h:96]: @njacazio @Barthelemy I am wondering if the Validity is a good name here, this feature limits the object creation time not the validity...

Comment (jgrosseo) [reply CCDB/include/CCDB/BasicCCDBManager.h:96]: Shall we just call it NotValidAfter as it is in the header?

Comment (shahor02) [reply CCDB/include/CCDB/BasicCCDBManager.h:96]: I think so.

Comment (njacazio) [reply CCDB/include/CCDB/BasicCCDBManager.h]: I agree with the change, I included it in last commit

Comment (chiarazampolli) [reply CCDB/include/CCDB/BasicCCDBManager.h]: Will you also add the CreatedNotBefore as suggested by @Barthelemy ?

**Comment by chiarazampolli** [comment]
> @njacazio for the line https://github.com/AliceO2Group/AliceO2/pull/3775/files#diff-3d82a935396c98479328a85f57c8c4a0R125: Why would you change it? The "time machine" sets the boundaries of the objects creation time by not their validity. Within these boundaries the you may still have different objects valid for different times, so the caching should work as usual. Have I misunderstood something?

I am not sure I understand the point here... 
The cache allows to sort of store the files, so that one does not need to go to CCDB all the time. Why, if you run in non-caching mode, you would not check the "CreatedNotAfter", if set? Nicolo' says it would have no effect, but why?

**Comment by shahor02** [comment]
@chiarazampolli I don't think caching and "TimeMachine" are interfering at all: caching means that when you ask for an object at some path (+ timestamp + whatever other conditions) for the 1st time it is stored locally in the BasicCCDBManager cache together with its uuid. Then, if you query for the same path again (with the same or other timestamps + conditions) AND provide the uuid of the object you have already cached, then no actual object transfer is done if the object with the same uuid would have been returned, otherwise new object and uuid is provided (and substitute old ones in the cache). And it should be irrelevant if one of the conditions in the query is `If_Not_After` etc. 

Just in case, pinging @costing for confirmation.

File: CCDB/include/CCDB/BasicCCDBManager.h (hunk 2 above)
Comment (njacazio) [reply CCDB/include/CCDB/BasicCCDBManager.h]: @chiarazampolli for this I would prefer opening a second PR if you agree

**Comment by njacazio** [comment]
@shahor02 @chiarazampolli  what I meant at line is that at line https://github.com/AliceO2Group/AliceO2/pull/3775/files#diff-3d82a935396c98479328a85f57c8c4a0R125 : 
```
    return mCCDBAccessor.retrieveFromTFileAny<T>(path, mMetaData, timestamp);
```
i.e. in case the CCDB manager is in "not cached" mode the object retrieval will not look for the `CreatedNotAfter` tag even if one has requested a limit in the object creation time.
I asked you because I found this behaviour misleading but if it's fine for you it's also fine for me.
Something like:
```
    return mCCDBAccessor.retrieveFromTFileAny<T>(path, mMetaData, timestamp, nullptr, "", mCreatedNotAfter ? std::to_string(mCreatedNotAfter) : "");
```
would have been more intuitive for me

**Comment by chiarazampolli** [comment]
> @njacazio for the line https://github.com/AliceO2Group/AliceO2/pull/3775/files#diff-3d82a935396c98479328a85f57c8c4a0R125: Why would you change it? The "time machine" sets the boundaries of the objects creation time by not their validity. Within these boundaries the you may still have different objects valid for different times, so the caching should work as usual. Have I misunderstood something?

I am not sure I understand the point here... 
The cache allows to sort of store the files, so that one does not need to go to CCDB all the time. Why, if you run in non-caching mode, you would not check the "CreatedNotAfter", if set? Nicolo' says it would have no effect, but why?

> @chiarazampolli I don't think caching and "TimeMachine" are interfering at all: caching means that when you ask for an object at some path (+ timestamp + whatever other conditions) for the 1st time it is stored locally in the BasicCCDBManager cache together with its uuid. Then, if you query for the same path again (with the same or other timestamps + conditions) AND provide the uuid of the object you have already cached, then no actual object transfer is done if the object with the same uuid would have been returned, otherwise new object and uuid is provided (and substitute old ones in the cache). And it should be irrelevant if one of the conditions in the query is `If_Not_After` etc.
> 
> Just in case, pinging @costing for confirmation.

I agree, I understand the same, but then should one not have line 125 with the check on the timeMachine condition too?

`return mCCDBAccessor.retrieveFromTFileAny<T>(path, mMetaData, timestamp, mCreatedNotAfter ? std::to_string(mCreatedNotAfter) : "");`

**Comment by chiarazampolli** [comment]
> @shahor02 @chiarazampolli what I meant at line is that at line https://github.com/AliceO2Group/AliceO2/pull/3775/files#diff-3d82a935396c98479328a85f57c8c4a0R125 :
> 
> ```
>     return mCCDBAccessor.retrieveFromTFileAny<T>(path, mMetaData, timestamp);
> ```
> 
> i.e. in case the CCDB manager is in "not cached" mode the object retrieval will not look for the `CreatedNotAfter` tag even if one has requested a limit in the object creation time.
> I asked you because I found this behaviour misleading but if it's fine for you it's also fine for me.
> Something like:
> 
> ```
>     return mCCDBAccessor.retrieveFromTFileAny<T>(path, mMetaData, timestamp, nullptr, "", mCreatedNotAfter ? std::to_string(mCreatedNotAfter) : "");
> ```
> 
> would have been more intuitive for me

Ciao,

As I just commented, I would also change it. I think that it should not matter if you are or not using the cache, the limit should still be used... But maybe I am missing something.

Chiara

**Comment by shahor02** [comment]
@njacazio ok, I've misunderstood your comment, sorry. Yes, the ``mCCDBAccessor.retrieveFromTFileAny`` calls with or w/o caching both should account for mCreatedNotAfter.
I hope the in case the ptr returned on https://github.com/AliceO2Group/AliceO2/blob/5d27b669180747bc9a3c03780a8e2325aab544e7/CCDB/include/CCDB/BasicCCDBManager.h#L128 is nullptr due to the CreatedNotAfter blocking otherwise matching object, the Error will be set in the headers.

**Comment by chiarazampolli** [comment]
> @chiarazampolli I don't think caching and "TimeMachine" are interfering at all: caching means that when you ask for an object at some path (+ timestamp + whatever other conditions) for the 1st time it is stored locally in the BasicCCDBManager cache together with its uuid. Then, if you query for the same path again (with the same or other timestamps + conditions) AND provide the uuid of the object you have already cached, then no actual object transfer is done if the object with the same uuid would have been returned, otherwise new object and uuid is provided (and substitute old ones in the cache). And it should be irrelevant if one of the conditions in the query is `If_Not_After` etc.
> 
> Just in case, pinging @costing for confirmation.

BTW, my understanding of the cache is that when you try to retrieve an object in the same path at given a timestamp, if the REST API finds that there is one with a different UUID, it returns it (https://github.com/AliceO2Group/AliceO2/blob/d931f9088e4b8b41f93d59c01b89604d56b57c41/CCDB/src/CcdbApi.cxx#L446), and I guess that then it returns the latest of all that do not match the UUID that you pass. I am saying it because it is a bit different from what @shahor02 writes (AND provide the uuid of the object): this UUID is not provided by the user, unless I misunderstand.

**Comment by chiarazampolli** [comment]
> @njacazio ok, I've misunderstood your comment, sorry. Yes, the `mCCDBAccessor.retrieveFromTFileAny` calls with or w/o caching both should account for mCreatedNotAfter.
> I hope the in case the ptr returned on
> 
> https://github.com/AliceO2Group/AliceO2/blob/5d27b669180747bc9a3c03780a8e2325aab544e7/CCDB/include/CCDB/BasicCCDBManager.h#L128
> is nullptr due to the CreatedNotAfter blocking otherwise matching object, the Error will be set in the headers.

This is also what I meant in my replies, so I agree.

**Comment by njacazio** [comment]
> @njacazio ok, I've misunderstood your comment, sorry. Yes, the `mCCDBAccessor.retrieveFromTFileAny` calls with or w/o caching both should account for mCreatedNotAfter.
> I hope the in case the ptr returned on
> 
> https://github.com/AliceO2Group/AliceO2/blob/5d27b669180747bc9a3c03780a8e2325aab544e7/CCDB/include/CCDB/BasicCCDBManager.h#L128
> is nullptr due to the CreatedNotAfter blocking otherwise matching object, the Error will be set in the headers.

@shahor02 I included my suggestion in last commit, concerning the Error tag in headers I don't know.
I can include a boost check on the header in the `testBasicCCDBManager.cxx` but the data member is private without access possibility so I'm afraid I don't know how to test it

File: CCDB/include/CCDB/BasicCCDBManager.h (hunk 2 above)
Comment (chiarazampolli) [reply CCDB/include/CCDB/BasicCCDBManager.h]: Sure, no problem for me, just let's not forget :-)

Comment (Barthelemy) [reply CCDB/include/CCDB/BasicCCDBManager.h]: I think that CreatedNotAfter is good. Also fine to make another PR for the counterpart method.

**Comment by shahor02** [comment]
>  this UUID is not provided by the user, unless I misunderstand.

We provide the cached uuid as a etag and the REST compares it with that of the object it would return.

**Comment by chiarazampolli** [comment]
> > this UUID is not provided by the user, unless I misunderstand.
> 
> We provide the cached uuid as a etag and the REST compares it with that of the object it would return.

What I meant was exactly that, while from your comment I thought you meant something else. Ok, we all know the same :-)

**Comment by jgrosseo** [comment]
Can we merge?

**Pull request merged**